File: src/nuitrack_bridge/types/hand.cc
#include "nuitrack_bridge/types/hand.h"
#include "nuitrack-rs/src/nuitrack_bridge/types/hand.rs.h"

namespace nuitrack_bridge::hand {


    std::shared_ptr<Hand> alias_hand_ptr(const std::shared_ptr<NuitrackHand>& sdk_hand_ptr) {
        if (!sdk_hand_ptr) {
            return nullptr;
        }

        NuitrackHand* raw_sdk_ptr = sdk_hand_ptr.get();
        Hand* casted_ptr = reinterpret_cast<Hand*>(raw_sdk_ptr);
        return std::shared_ptr<Hand>(sdk_hand_ptr, casted_ptr);
    }

    // --- UserHands functions ---
    int32_t getUserHandsUserId(const UserHands& userHands) {
        return static_cast<int32_t>(userHands.userId);
    }

    std::shared_ptr<Hand> getUserHandsLeftHand(const UserHands& userHands) {
        return alias_hand_ptr(userHands.leftHand);
    }

    std::shared_ptr<Hand> getUserHandsRightHand(const UserHands& userHands) {
        return alias_hand_ptr(userHands.rightHand);
    }
} 

File: src/nuitrack_bridge/types/hand.rs
#[cxx::bridge(namespace = "nuitrack_bridge::hand")]
pub mod ffi {

    #[derive(Debug, Clone, Copy, PartialEq)]
    pub struct Hand {
        pub x: f32,
        pub y: f32,
        pub click: bool,
        pub pressure: i32,
        pub x_real: f32,
        pub y_real: f32,
        pub z_real: f32,
    }
    
    #[namespace = "tdv::nuitrack"]
    unsafe extern "C++" {
        type UserHands;
    }

    impl CxxVector<UserHands> {}

    unsafe extern "C++" {
        include!("nuitrack_bridge/types/hand.h");

        // --- NuitrackUserHands accessors ---
        #[cxx_name = "getUserHandsUserId"] // Corrected
        pub fn user_hands_user_id(user_hands: &UserHands) -> i32;
        
        #[cxx_name = "getUserHandsLeftHand"] // Corrected
        pub fn user_hands_left_hand(user_hands: &UserHands) -> SharedPtr<Hand>;
        #[cxx_name = "getUserHandsRightHand"] // Corrected
        pub fn user_hands_right_hand(user_hands: &UserHands) -> SharedPtr<Hand>;
    }
}